LFS-159: As a user I can access a dashboard

acrowthe opened the pull request:
Replace userdashboard with extension based grid
- Reduces number of + buttons
- Allows access to subjects from homepage
- Allows extensibility on a per-project basis
    
Default displayed items:
- Forms list, with tabs for completed forms and draft forms
- Subject list, with tabs for subject type
    
Default creatable items:
- Forms
- Subjects

Reuses existing components for creating forms and subjects
- Added ability to specify component from an extensionRenderURL, so NewSubjectDialog could be used instead of the SubjectSelector default
- Adjusted NewFormDialog to add a dialog only mode

acrowthe committed:
LFS-159: As a user I can access a dashboard
Replace userdashboard with extension based grid
- Reduces number of + buttons
- Allows access to subjects from homepage
- Allows extensibility on a per-project basis

Default displayed items:
- Forms list, with tabs for completed forms and draft forms
- Subject list, with tabs for subject type

Default creatable items:
- Forms
- Subjects

acrowthe committed:
LFS-159: As a user I can access a dashboard
- Improve new item dialog to a dynamic table

acrowthe committed:
LFS-159: As a user I can access a dashboard
- Fix non-functional cancel button on NewFormDialog
- Remove unneeded material table title from add item dialog

acrowthe commented:
I'm not sure what would be the best text for the dialog title and add button tooltip: Suggestions would be appreciated
![image](https://user-images.githubusercontent.com/20056751/101658554-208f8b80-3a13-11eb-9e84-e99a9bd327ea.png)
![image](https://user-images.githubusercontent.com/20056751/101658558-22594f00-3a13-11eb-9c0b-16acf46bd007.png)

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/SubjectView.jsx:
```diff
@@ -0,0 +1,128 @@
+
+  let fetchSubjectTypes = () => {
+    let url = new URL("/query", window.location.origin);
+    url.searchParams.set("query", `SELECT * FROM [lfs:SubjectType] as n`);
```
Once #380 is merged, please order the subject types:
```suggestion
    url.searchParams.set("query", `SELECT * FROM [lfs:SubjectType] as n order by n.'lfs:defaultOrder'`);
```

marta- replied:
#380 is merged, please rebase and update the tab order

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/UserDashboard.jsx:
```diff
@@ -16,163 +16,164 @@
       </Grid>
+      <Dialog open={open} onClose={() => setOpen(false)}>
+        <DialogTitle>
+          <Typography variant="h6" className={classes.dialogTitle}>Add</Typography>
```
```suggestion
          <Typography variant="h6" className={classes.dialogTitle}>New</Typography>
```

marta- replied:
`DialogTitle` already has a `Typography` component into which it renders the provided title unless `disableTypography` is specified. This generates a console warning `Warning: validateDOMNesting(...): <h6> cannot appear as a child of <h2>`. If custom styling of the title isn't a must, just passing the text should suffice.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/UserDashboard.jsx:
```diff
@@ -16,163 +16,164 @@
+        </DialogActions>
+      </Dialog>
+      <div className={classes.newFormButtonWrapper}>
+        <Tooltip title={"Add"} aria-label="add">
```
"Add" is fine, "New" would also work and be consistent with the "New subject", "New form" tooltips.
```suggestion
        <Tooltip title={"New"} aria-label="add">
```

marta- reviewed (commented):
Looks great, just a few initial comments, more to come later.

sashaandjic commented:
The new look is fantastic. May I suggest condensing grids? The row appear a bit too high.
![image](https://user-images.githubusercontent.com/23243244/101666849-9fd58d00-3a1c-11eb-90de-2ff6dbc51c39.png)

sashaandjic commented:
As for the Add dialog, do we even need title, tyoe, pagination and search at all?

marta- commented:
> As for the Add dialog, do we even need title, tyoe, pagination and search at all?

The dialog is using the same widget as the other steps of creating a new entity, so it's good to be consistent. However, this is more likely than others to have fewer entries, our current projects may only add 1 or 2 extra entities.

I would suggest to only showing pagination if there's more than one page, and removing the table headers ("Type"). @acrowthe here's a quick example of how to deal with the pagination: https://github.com/ccmbioinfo/lfs/pull/354/commits/b28b0a1538bce798ee96eb0a6e7855fc204a931d#diff-8abd711cb3808388093456509d843732e1ba62974461644f36d718ac320ff89fR385-R388

I would keep the search and the title, for consistency with the next steps.

marta- commented:
> May I suggest condensing grids? The row appear a bit too high.

I think this is the default look of our livetable. I wouldn't change it yet, as there's a chance we may swap it for a virtualized list later on, tbd. For now I would say it works as is.

> I am also curious whether doing the left alignment, rather than center, for the tab labels would improve the look.

I will make a mockup soon with some improvements for the card headers, which should help with the alignment.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/FormView.jsx:
```diff
@@ -0,0 +1,84 @@
+      />
+      <Tabs value={activeTab} onChange={(event, value) => setActiveTab(value)}>
+        {tabs.map(value => {
+          return <Tab label={value} />;
```
Needs a `key` property (this currently generates a JS console warning).

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/SubjectView.jsx:
```diff
@@ -0,0 +1,128 @@
+          ? <CircularProgress/>
+          : <Tabs value={activeTab} onChange={(event, value) => setActiveTab(value)}>
+              {subjectTypes.map(subject => {
+                return <Tab label={subject['label'] || subject['@name']} />;
```
Needs a `key` property (this currently generates a JS console warning).

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/UserDashboard.jsx:
```diff
@@ -16,163 +16,164 @@
+        {
+          dashboardExtensions.map((extension) => {
+            let Extension = extension["lfs:extensionRender"];
+            return <Grid item lg={12} xl={6}>
```
Needs a `key` property (this currently generates a JS console warning).

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/UserDashboard.jsx:
```diff
@@ -16,163 +16,164 @@
+      {
+        creationExtensions.map((extension, index) => {
+          let Extension = extension["lfs:extensionRender"];
+          return <Extension
```
Needs a `key` property (this currently generates a JS console warning).

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/UserDashboard.jsx:
```diff
@@ -16,163 +16,164 @@
+            data={creationExtensions}
+            options={{
+              search: true,
+              pageSize: rowCount,
```
If all entries will be shown on the first (and only) page, pagination should be completely removed.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/UserDashboard.jsx:
```diff
@@ -16,163 +16,164 @@
+        </DialogTitle>
+        <DialogContent>
+          <MaterialTable
+            columns={[
+              { title: 'Type', field: 'lfs:extensionName' },
+            ]}
```
Since there's only one column, remove the header row.

acrowthe committed:
LFS-159: As a user I can access a dashboard
Resolve code review comments and console errors

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/UserDashboard.jsx:
```diff
@@ -16,163 +16,172 @@
+              search: true,
+              pageSize: rowCount,
+              showTitle: false,
+              paging: creationExtensions.length > 5,
```
Shouldn't it be
```suggestion
              paging: creationExtensions.length > rowCount,
```
?

acrowthe replied:
I thought about doing that, but was worried about the user losing the ability to lower the rowCount if they set it too high.
Eg. 19 entries, set rowCount to 20 and then having pagination and the option to change rowcount hidden

marta- replied:
I see, that's fine then.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/UserDashboard.jsx:
```diff
@@ -91,28 +96,27 @@ function UserDashboard(props) {
           <MaterialTable
             columns={[
-              { title: 'Type', field: 'lfs:extensionName' },
+              { field: 'lfs:extensionName' },
```
Looks like this generates an empty header. Please hide the header in options.

marta- replied:
This is what the dialog would look like if the header and the search (toolbar) are hidden. I think that's nice and clean, and @sashaandjic was right about not needing search. let's hide it.

![image](https://user-images.githubusercontent.com/651980/101693520-b93c0080-3a3f-11eb-8d1d-6be26a3834a3.png)

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/SubjectView.jsx:
```diff
@@ -0,0 +1,130 @@
+              columns={columns}
+              customUrl={'/Subjects.paginate?fieldname=type&fieldvalue='+ encodeURIComponent(subjectTypes[activeTab]["jcr:uuid"])}
+              defaultLimit={10}
+              filters
```
No filters here, they don't work as expected on subjects

```suggestion
```

marta- committed:
LFS-159: As a user I can access a dashboard
Simplified the resource creation dialog

marta- committed:
LFS-159: As a user I can access a dashboard
Removed non-functional filters from the SubjectView box and ensured that the two boxes have the same hight regardsless of their content

marta- committed:
LFS-159: As a user I can access a dashboard
Embelished dashboard boxes:
* added avatars
* re-styled and aligned tabs
* aligned and restyled filters
* addied link to expanded view of Subjects/Forms
* added Livetable options to not display top or bottom pagination, hid top pagination in dashboard boxes, tidied up table spacing

acrowthe commented:
Known visual bug:
When the windows is half of a 1080p screen wide and there is data in the forms table, switching tabs from completed to draft or vice versa will cause the forms table to move from below subjects to beside subjects and back.

Should be logged as an issue if merged as is, but not a show stopper

sdumitriu commented on modules/commons/src/main/frontend/src/assetManager.jsx:
```diff
@@ -61,10 +61,35 @@ var getAssetURL = async function(assetUrl) {
+//
+// @param {string} assetURL the URL to extract the parameters from
+// @return an object containing the parameters as key value pairs
+var getURLParameters = (assetUrl) => {
```
You shouldn't do manual parsing, use [`URLSearchParams`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams).

sdumitriu commented on modules/data-entry/src/main/resources/SLING-INF/content/Extensions.json:
```diff
@@ -83,5 +83,39 @@
+            "lfs:defaultOrder": 1
+        }
+    },
+    "DashboardCreation": {
```
They should all be singular or plural, mixing is not good. I would rename this `DashboardCreations`.

Still, Creations bothers me, it doesn't sound right, but I can't find an obviously better alternative. Maybe `DashboardCreateMenuEntries`? Any other suggestions?

acrowthe replied:
Changed to DashboardMenuItems.

sdumitriu commented on modules/commons/src/main/frontend/src/assetManager.jsx:
```diff
@@ -85,7 +111,8 @@ var loadAsset = async function(assetURL) {
             .then(remoteComponentSrc => {
               var returnVal = window.eval(remoteComponentSrc);
-              return returnVal.default;
+              return parameters?.component ? returnVal[parameters.component] : returnVal.default;
```
:+1:

sashaandjic commented:
The new dashboard looks **fantastic**. One small suggestion: instead of  "see more" text in tooltip, perhaps "see all" would be more user friendly?
![image](https://user-images.githubusercontent.com/23243244/101802023-f904f580-3adc-11eb-823b-7258cfed464a.png)

marta- commented:
> One small suggestion: instead of "see more" text in tooltip, perhaps "see all" would be more user friendly?

"See more" is not great, but "See all" is inaccurate as it's exactly the same resources in a wider table with more columns. "More"? "Expand"?

> Question: With the new dashboard look, do we need Patient and Tumors buttons in the left panel?

We don't. Tabs will be added to Subjects and Forms, and Patients/Tumors will be removed.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/FormView.jsx:
```diff
@@ -0,0 +1,107 @@
+        avatar={<Avatar className={classes.formViewAvatar}><DescriptionIcon/></Avatar>}
+        title={<Typography variant="h6">Forms</Typography>}
+        action={
+          <Tooltip title="See more">
```
```suggestion
          <Tooltip title="Expand">
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/SubjectView.jsx:
```diff
@@ -0,0 +1,141 @@
+        avatar={<Avatar className={classes.subjectViewAvatar}><AssignmentIndIcon/></Avatar>}
+        title={<Typography variant="h6">Subjects</Typography>}
+        action={
+          <Tooltip title="See more">
```
```suggestion
          <Tooltip title="Expand">
```

acrowthe committed:
LFS-159: As a user I can access a dashboard
Address code review comments
- Rename DashboardCreation to DashboardMenuItems
- Switch from manual asset URL parsing to URLSearchParams
- Minor text changes

sashaandjic commented:
Once the filter is set in the dashboard, it is persistent when moving from Completed and Draft forms. When I press on the Expand button I would also expect to see the filtered forms, however the filter is lost at that point.
![image](https://user-images.githubusercontent.com/23243244/101821324-d206ed80-3af5-11eb-9080-c2595c3e374e.png)

sashaandjic commented:
Expand button has inconsistent behaviour for subjects and forms.
For SUBJECTS it shows **all types** of subjects
For FORMS it shows **only completed** forms

acrowthe commented:
> Expand button has inconsistent behaviour for subjects and forms.
For SUBJECTS it shows all types of subjects
For FORMS it shows only completed forms

The expand button just opens the Subjects or Forms page. Tabs will get added to those pages in a future PR to add similar functionality to those pages as seen in the dashboard (list by subject type or completion status)

marta- commented:
> The expand button just opens the Subjects or Forms page. Tabs will get added to those pages in a future PR to add similar functionality to those pages as seen in the dashboard (list by subject type or completion status)

FTR, this is the task: https://phenotips.atlassian.net/browse/LFS-780

sashaandjic reviewed (approved):
tested

marta- committed:
LFS-159: As a user I can access a dashboard
Renamed "User dashboard" to "Dashboard"

sdumitriu reviewed (changes requested):
Bug: create a new Demographics form, set gender = M, save, go to dashboard, add a filter for Gender = M, JSON parsing error. Looks like a problem with the query.

Seems that the only filters that still work are for Subject and Questionnaire.

sdumitriu committed:
LFS-159: As a user I can access a dashboard
Fixed broken filters

sdumitriu reviewed (approved):
Filters fixed, everything else looks good.

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/UserDashboard.jsx:
```diff
@@ -16,163 +16,172 @@
-      })
-      .catch(handleError);
+  let onClose = () => {
+    setSelectedCreation(-1);
```
This causes a JS error about updating unmounted components after creating a new form with a new subject.
